[EXPERIMENTAL][DI] Add tail injection for methods + autowiring

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

ping @fabpot @simensen

```yaml
services:
    foo:
        class: Foo
        tails:
            method1: { 1: 'bar', 2: '@bar' }
```

will create an inheritance proxy overriding `Foo::method1`, making its args 2 & 3 optional, with the configured values being used as defaults, keeping only argument 1 mandatory (if it was in the original declaration of course).

Allows Laravel like controllers (`function indexAction(Request $request, PostRepository $postRepository)` with `$postRepository` injection), but with full Symfony-style flexibility (ie you can/have to exactly choose where+what to inject).
Note that this can *not* be achieved via a 3.2 `ArgumentResolver` : only compile time can achieve the required level of configurability - not runtime, which `ArgumentResolver` is only about.
Works with any service really of course.
Reuses the inheritance proxy generation already in place for getter injection.

nicolas-grekas committed:
[EXPERIMENTAL][DI] Add tail injection for methods + autowiring

nicolas-grekas commented:
Note that autowiring will be added later - the PR is big enough.
The way I see it requires https://github.com/symfony/symfony/pull/21194
And declaration in Yaml would look like:
```yaml
services:
  App\Controller\FooController:
    autowire:
      - !tail '*Action'
```

stof commented:
this requires that the arguments are at the end, right ?

nicolas-grekas commented:
yes, it throws if there are remaining required tail args.

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1325,40 +1335,161 @@ private function generateOverriddenGetters($id, Definition $definition, \Reflect
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" cannot be marked as final.', $msg, $id, $class->name, $r->name));
         }
+        if ($type = method_exists($r, 'getReturnType') ? $r->getReturnType() : null) {
+            $type = ': '.($type->allowsNull() ? '?' : '').self::generateTypeHint($type, $r);
```
I vote for moving the building of the dumped return type outside this method. I already found the reference argument harder to read previously, but it becomes worse when nesting the resolution by reference in another method. We already return all the necessary info anyway, as this logic depends only on the reflector anyway

nicolas-grekas replied:
done

stof commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -1488,4 +1622,17 @@ private function getExpressionLanguage()
+            $code[] = sprintf('%s => %s', var_export($k, true), self::export($v));
+        }
+
+        return sprintf('array(%s)', implode(', ', $code));
```
isn't it exactly what ``var_export($array)`` does ?

nicolas-grekas replied:
with one difference: this generates a single line export, while var_export splits arrays on several lines

nicolas-grekas commented:
PR is ready, with tests and autowiring.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ChildDefinition.php:
```diff
@@ -162,11 +162,21 @@ public function setDeprecated($boolean = true, $template = null)
      * {@inheritdoc}
      */
-    public function setAutowiredMethods(array $autowiredMethods)
+    public function setAutowiredCalls(array $autowiredCalls)
```
@dunglas @stof if you're ok with this rename, I can submit a refactoring PR first, to reduce the diff of this PR, OK for you?

stof replied:
looks good to me. And doing it now is still possible as the existing name is not yet released

nicolas-grekas commented:
"partial" renamed to "tail"

stof commented on src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:
```diff
@@ -33,6 +33,7 @@ class AutowirePass extends AbstractRecursivePass
      * @internal
      */
     const MODE_OPTIONAL = 2;
+    const MODE_TAIL = 3;
```
should also be ``@internal``

stof commented:
@nicolas-grekas why is this PR opened on your fork rather than on the symfony repo ? this looks weird

lyrixx committed:
[Bridge/Monolog] Enhanced the Console Handler

Basically, the formatter now uses the VarDumper & uses more significant
colors and has a more compact / readable format (IMHO).

nicolas-grekas commented:
@stof because this is going to be even more surprising than getter injection to some :)
I'd like the idea to be confirmed first with a small core-team related group.
What do you think yourself of the usefulness of this?

fabpot committed:
fixed CS

nicolas-grekas committed:
fix typo

nicolas-grekas committed:
[DI] Replace PHP7-conditional return-type checks by regular type-hints that work on PHP5

nicolas-grekas committed:
minor #21888 [DI] Replace PHP7-conditional return-type checks by regular type-hints that work on PHP5 (nicolas-grekas)

This PR was merged into the 3.3-dev branch.

Discussion
----------

[DI] Replace PHP7-conditional return-type checks by regular type-hints that work on PHP5

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Commits
-------

372ff7c [DI] Replace PHP7-conditional return-type checks by regular type-hints that work on PHP5

javiereguiluz committed:
Added a castToArray() config helper

wouterj committed:
Count @expectedDeprecation as an assertion

fabpot committed:
bug #21896 [PHPunitBridge] Count @expectedDeprecation as an assertion (wouterj)

This PR was merged into the 3.2 branch.

Discussion
----------

[PHPunitBridge] Count @expectedDeprecation as an assertion

| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

By adding `addToAssertionCount()`, the `@expectedDeprecation` annotation is recognized as an assertion by PHPUnit. This means PHPUnit will not report the test as risky because it does not contain any assertion.

Commits
-------

ba5c0f4a47 Count @expectedDeprecation as an assertion

fabpot committed:
Merge branch '3.2'

* 3.2:
  Count @expectedDeprecation as an assertion

fabpot committed:
feature #21893 Added a castToArray() config helper (javiereguiluz)

This PR was squashed before being merged into the 3.3-dev branch (closes #21893).

Discussion
----------

Added a castToArray() config helper

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21817
| License       | MIT
| Doc PR        | -

Commits
-------

3a589df5cb Added a castToArray() config helper

fabpot committed:
feature #21705 [Bridge/Monolog] Enhance the Console Handler (lyrixx)

This PR was merged into the 3.3-dev branch.

Discussion
----------

[Bridge/Monolog] Enhance the Console Handler

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | -
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

---

I extracted and enhanced the formatting logic from #21080.
Basically, The formatter now use the VarDumper & use more significant colors and has a more compact / readable format (IMHO).

---

I used the following code to generate before/after screenshots.

```php
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $logger = $this->getContainer()->get('logger');
        $filesystem = $this->getContainer()->get('filesystem');

        $anObject = new \stdClass;
        $anObject->firstPpt = 'foo';
        $anObject->secondePpt = 'bar';

        $logger->log('notice', 'Hello {who}', [
            'who' => 'Wold',
            'an_object' => $anObject,
            'file_system' => $filesystem,
        ]);

        $r = new \ReflectionClass(LogLevel::class);
        foreach ($r->getConstants() as $level) {
            $logger = $logger->withName($level);
            $logger->log($level, 'log at level {level}', [
                'level' => $level,
            ]);
        }
    }
```

before:

![screenshot12](https://cloud.githubusercontent.com/assets/408368/23183075/0bb21f80-f87b-11e6-8123-f6da70f2493b.png)

After:

![screenshot11](https://cloud.githubusercontent.com/assets/408368/23182971/b4022ed8-f87a-11e6-9d3b-de1a9d4ce9aa.png)

Commits
-------

b663ab5246 [Bridge/Monolog] Enhanced the Console Handler
